### ISSUE TITLE
Most tests do not verify assumptions of the method

### ISSUE THREAD

**Issue by jorgecarleitao** [shap/shap#469]
A brief look at the tests, e.g. [test_deep](https://github.com/slundberg/shap/blob/fce3f2f4a706c33f7280bb91e8c6c1efb55b5702/tests/explainers/test_deep.py#L87), [test_linear](https://github.com/slundberg/shap/blob/fce3f2f4a706c33f7280bb91e8c6c1efb55b5702/tests/explainers/test_linear.py#L39), [test_tree](https://github.com/slundberg/shap/blob/fce3f2f4a706c33f7280bb91e8c6c1efb55b5702/tests/explainers/test_tree.py#L194), [test_gradient](https://github.com/slundberg/shap/blob/fce3f2f4a706c33f7280bb91e8c6c1efb55b5702/tests/explainers/test_gradient.py#L86), shows that we are either testing that the sum of shap values is the expected value, or that the code runs.

In other words, we almost never check that the code actually does what it is supposed to do in terms of the individual shap values. I understand that it is not easy to design tests for this, but it is part of the research and development to find suitable test suits for this problem.

In PR #468 I added a test to verify what I understood from the documentation of LinearExplainer. Tests for other explainers are necessary to guarantee that the explanations are correct in controlled data (and thus expected to be correct in uncontrolled data).

**Comment by slundberg** [comment]
Thanks for the new test! I should clarify that we typically test that the sum of the SHAP values equals the output of the model for each prediction. While it is possible for that to hold while the algorithm itself fails, it is very unlikely. We also have comparison tests between different methods when applicable. At some point I had brute force tests for lots of Tree SHAP explanations, but it looks like those are not in the unit tests right now, perhaps they should get added back in. Of course I hope to continue to add new unit tests, and welcome PRs that add more :)
